pallets/xcm/src/lib.rs

```diff
@@ -36,16 +40,68 @@ pub mod pallet {
 
 	/// The module configuration trait.
 	#[pallet::config]
-	pub trait Config: frame_system::Config {}
+	pub trait Config: frame_system::Config {
+		/// The overarching event type.
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+
+		type XcmExecutor: ExecuteXcm<Self::Call>;
+
+		#[pallet::constant]
+		type MaxWeight: Get<Weight>;
+	}
 
 	#[pallet::error]
-	pub enum Error<T> {}
+	pub enum Error<T> {
+	}
 
 	#[pallet::hooks]
 	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {}
 
 	#[pallet::call]
 	impl<T: Config> Pallet<T> {}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	#[pallet::metadata(T::BlockNumber = "BlockNumber")]
+	pub enum Event<T: Config> {
+		/// Downward message is invalid XCM.
+		/// \[ id \]
+		InvalidFormat([u8; 8]),
+		/// Downward message is unsupported version of XCM.
+		/// \[ id \]
+		UnsupportedVersion([u8; 8]),
+		/// Downward message executed with the given outcome.
+		/// \[ id, outcome \]
+		ExecutedDownward([u8; 8], Outcome),
+	}
+
+	/// For an incoming downward message, this just adapts an XCM executor and executes DMP messages
+	/// immediately up until some `MaxWeight` at which point it errors. Their origin is asserted to be
+	/// the Parent location.
+	impl<T: Config> DownwardMessageHandler for Pallet<T> {
+		fn handle_downward_message(msg: InboundDownwardMessage) -> Weight {
+			let id = sp_io::hashing::twox_64(&msg.msg[..]);
+			let msg = VersionedXcm::<T::Call>::decode(&mut &msg.msg[..])
+				.map(Xcm::<T::Call>::try_from);
```

Comment:
the weight for this is left out intentionally, I assume?

Reply:
correct. the weight returned is the unknowable weight. stuff for bookkeeping, dispatch, decode and conversion can be benchmarked normally.